### PR TITLE
Dynamic selection of the persistence implementation

### PR DESCRIPTION
**Pull request by np**

### CODE REVIEW

**Comment by fisx** [comment]
btw would it make sense to have three backend options, namely AcidStateMem, AcidStateDisk, STM, and drop the `*InMemory` variants of `mkPersist` etc.?  i think it would.

File: src/Persistent/Implementation.hs [hunk 1]
```diff
@@ -1,16 +1,22 @@
...
+
+withPersist' :: Config -> (forall r. (PersistM r, GenArbitrary r) => RunPersistNat IO r -> IO a) -> IO a
+withPersist' cfg = withPersist (mkRunPersist cfg)
+
+mkRunPersist :: Config -> IO RunPersist
+mkRunPersist cfg =
+    case cfg ^. persistenceImpl of
+        STM             -> mkRunPersistSTM
+        AcidStateInMem  -> mkRunPersistInMemory
+        AcidStateOnDisk -> mkRunPersistOnDisk cfg
```
Comment (np) [inline src/Persistent/Implementation.hs]: @fisx isn't that what you were looking for?

Comment (fisx) [inline src/Persistent/Implementation.hs]: yes, that looks nice!

**Comment by np** [comment]
I'd like to have this merged regardless of #328

**Comment by Mikolaj** [comment]
A good stuff.

A note: with this setup we have to remember to never run tests with `AcidStateOnDisk` (unless we start all individual tests by wiping out the DB, which I did do in some of my old reverted commits and they did pass with `AcidStateOnDisk`), but only with any of the other two. In the previous idiom, I hardcoded that tests are always run with mkRunPersistenInMemory, so there was no possibility of a mistake. Also, we don't want want to run tests on disk so as not to corrupt a persistent DB (production, staging, demo or any other) via tests. Anyway, just a warning to keep in mind.

File: tests/Persistent/ApiSpec.hs [hunk 2]
```diff
@@ -148,55 +148,63 @@ changeAUID (AUID i) = AUID (succ i)
...
-    addDbSpec "addIdea"  getIdeas addIdea
-    getDbSpec "getWildIdeas"      getWildIdeas
-    getDbSpec "getIdeasWithTopic" getIdeasWithTopic
+    context "Using STM"             $ persistApiSpec STM
+    context "Using AcidStateInMem"  $ persistApiSpec AcidStateInMem
+
+    -- I guess we need to change the config to target a different directory.
+    -- And wipe out the data every time.
+    -- context "Using AcidStateOnDisk" $ persistApiSpec AcidStateOnDisk
+
```
Comment (Mikolaj) [inline tests/Persistent/ApiSpec.hs]: yep, see my top-level comment on the PR

**Pull request merged**